### ISSUE TITLE
JavaScript built-in objects are different in each module

### ISSUE THREAD

**Issue by wiktortomczak** [laurentj/slimerjs#611]
Built-in JS objects such as `Object`, `Error`, `Uint8Array` (constructors) appear to have different identity in different modules imported via `require()`. Specifically, if module `a.js` passes a reference to eg. `Object` to function `f()` imported from module `b.js`, the `Object` in `b.js` is different from the `Object` received from `a.js`.

See **Steps to reproduce** below for more details.

https://docs.slimerjs.org/current/api/js-standard.html says: "Most Javascript built-in objects are available and can be used as expected". This is not as expected.

### versions 

- SlimerJS: 0.10.3
- Firefox: 52.0.2
- Operating system: Ubuntu 16.04

### Steps to reproduce the issue

`a.js`

```javascript
var b = require('./b');

b.f(Object, Error, Uint8Array);
```

`b.js`

```javascript
exports.f = function(object, error, uint8array) {
  console.log(object === Object);
  console.log(error === Error);
  console.log(uint8array === Uint8Array);

  console.log(Object === Object);
  console.log(Error === Error);
  console.log(Uint8Array === Uint8Array);
};
```

```
$ slimerjs a.js
```

### Actual results:

```
false 
false 
false 
true 
true 
true 
```

### Expected results:

```
true
true
true
true 
true 
true 
```

**Note**: One workaround is to assign specific global objects to `require.globals` before calling `require()`:

`a.js` with the workaround:

```javascript
require.globals.Object = Object;
require.globals.Error = Error;
require.globals.Uint8Array = Uint8Array;

var b = require('./b');

b.f(Object, Error, Uint8Array);
```

```
$ slimerjs a.js
true
true
true
true 
true 
true 
```

**Comment by laurentj** [comment]
>https://docs.slimerjs.org/current/api/js-standard.html says: "Most Javascript built-in objects are available and can be used as expected". This is not as expected.

No. This is as expected. You can use any JS object as expected. You have all properties, methods etc...

The problem that it appears here, is that it seems there is a copy of objects passed to other modules, not a "reference". Each modules don't run in the same JS compartment/sandbox, so this could explain why. 

I should verify if I don't do mistakes with JS sandboxes of SpiderMonkey (the javascript engine). Probably it's a "normal" behavior in SpiderMonkey (and if so, I can't do anything).

**Comment by wiktortomczak** [comment]
Thanks for a quick reply.

I think it's reasonable to expect that Object === Object works. This is how it works in a browser. JS libraries that I use rely on that: https://github.com/google/protobuf/blob/f873d32/js/binary/utils.js#L964.

I'm suspecting the same cause - copy of objects resulting from how module sandboxing is implemented.

**Comment by laurentj** [comment]
>This is how it works in a browser.

This is how it works in a **web page**, not in a browser (as browsers don't support ES6 modules yet) ;-)

And here, you are not in a web page, you are inside a browser ;-)

But ok, I understand you expect Object === Object. I'll verify that.

**Comment by wiktortomczak** [comment]
Thanks!

**Comment by RReverser** [comment]
> This is how it works in a browser.

Not really. In browser you also can have different realms, except they're usually constructed by iframes. In this case, SlimerJS adds new thing - modules - that browsers don't have natively yet, and each module also has its own realm like iframe.
